modules/permutiveRtdProvider.js

```diff
@@ -383,17 +377,57 @@ function iabSegmentId(permutiveSegmentId, iabIds) {
   return iabIds[permutiveSegmentId] || unknownIabSegmentId
 }
 
+/**
+ * Pull the latest configuration and cohort information and update accordingly.
+ *
+ * @param reqBidsConfigObj - Bidder provided config for request
+ * @param customModuleConfig - Publisher provide config
+ */
+function readAndSetCohorts(reqBidsConfigObj, moduleConfig) {
+  const segmentData = getSegments(deepAccess(moduleConfig, 'params.maxSegs'))
+
+  makeSafe(function () {
+    // Legacy route with custom parameters
+    // ACK policy violation, in process of removing
+    setSegments(reqBidsConfigObj, moduleConfig, segmentData)
+  });
+
+  makeSafe(function () {
+    // Route for bidders supporting ORTB2
+    setBidderRtb(reqBidsConfigObj.ortb2Fragments?.bidder, moduleConfig, segmentData)
+  })
+}
+
+let permutiveReadyCallbackRegistered = false
+
 /** @type {RtdSubmodule} */
 export const permutiveSubmodule = {
   name: MODULE_NAME,
   getBidRequestData: function (reqBidsConfigObj, callback, customModuleConfig) {
+    const completeBidRequestData = () => {
+      logger.logInfo(`Request data updated`)
+      callback()
+    }
+
+    const moduleConfig = getModuleConfig(customModuleConfig)
+
+    readAndSetCohorts(reqBidsConfigObj, moduleConfig)
+
     makeSafe(function () {
-      // Legacy route with custom parameters
-      initSegments(reqBidsConfigObj, callback, customModuleConfig)
-    });
-    makeSafe(function () {
-      // Route for bidders supporting ORTB2
-      setBidderRtb(reqBidsConfigObj.ortb2Fragments?.bidder, customModuleConfig)
+      if (!permutiveReadyCallbackRegistered && moduleConfig.waitForIt && isPermutiveOnPage()) {
+        // Prevent multiple calls to set cohorts
+        permutiveReadyCallbackRegistered = true
+
+        window.permutive.ready(function () {
```

Comment:
Curious what difference this makes compared to previously? Is it only for updating cohorts?

Reply:
I refactored this as the previously existing `initSegments` function would actually call `callback()` before the module updated the ortb2 interface. I'm not sure what the implications of that are, but the `callback` should be called once everything is finished in the RTD module according to the [docs](https://docs.prebid.org/dev-docs/add-rtd-submodule.html#getbidrequestdata).

Another issue I noted was that if the publisher had `config.waitForIt`, it would set the legacy segments twice on each call as it would keep registering a callback even after the Permutive SDK is in real-time. When this happens, the callback is executed immediately from then onwards, performing a double call. 
This now only registers through the callback until Permutive is in real-time and then once it is in realtime, just completes the RTD auction as usual.